hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/StreamBlockInput.java

```diff
@@ -0,0 +1,777 @@
+    // and is retried, we need the previous position.  Position is reset after
+    // successful read in adjustBufferPosition()
+    blockPosition = getPos();
+    bufferOffsetWrtBlockDataData = readData(startByteIndex, len);
+    long tempOffset = 0L;
+    buffersSize = 0L;
+    bufferoffsets = new ArrayList<>(buffers.size());
+    for (ByteBuffer buffer : buffers) {
+      bufferoffsets.add(tempOffset);
+      tempOffset += buffer.limit();
+      buffersSize += buffer.limit();
+
+    }
+    bufferIndex = 0;
+    allocated = true;
+    adjustBufferPosition(startByteIndex - bufferOffsetWrtBlockDataData);
+
+  }
+
+  @VisibleForTesting
+  protected long readData(long startByteIndex, long len)
+      throws IOException {
+    Pipeline pipeline = pipelineRef.get();
+    buffers = new ArrayList<>();
+    ReadBlockResponseProto response =
+        ContainerProtocolCalls.readBlock(xceiverClient, startByteIndex,
+        len, blockID, validators, tokenRef.get(), pipeline.getReplicaIndexes(), verifyChecksum);
+    List<ReadChunkResponseProto> readBlocks = response.getReadChunkList();
+
+    for (ReadChunkResponseProto readBlock : readBlocks) {
+      if (readBlock.hasDataBuffers()) {
+        buffers.addAll(BufferUtils.getReadOnlyByteBuffers(
+            readBlock.getDataBuffers().getBuffersList()));
+      } else {
+        throw new IOException("Unexpected error while reading chunk data " +
+            "from container. No data returned.");
+      }
+    }
+    return response.getReadChunk(0)
+        .getChunkData().getOffset();
+  }
+
+  /**
+   * Check if the buffers have any data remaining between the current
+   * position and the limit.
+   */
+  private boolean buffersHaveData() {
+    boolean hasData = false;
+    if (buffersAllocated()) {
+      int buffersLen = buffers.size();
+      while (bufferIndex < buffersLen) {
+        ByteBuffer buffer = buffers.get(bufferIndex);
+        if (buffer != null && buffer.hasRemaining()) {
+          // current buffer has data
+          hasData = true;
+          break;
+        } else {
+          if (bufferIndex < buffersLen - 1) {
+            // move to next available buffer
+            ++bufferIndex;
+            Preconditions.checkState(bufferIndex < buffers.size());
+          } else {
+            // no more buffers remaining
+            break;
+          }
+        }
+      }
+    }
+
+    return hasData;
+  }
+
+  /**
+   * Check if there is more data in the chunk which has not yet been read
+   * into the buffers.
+   */
+  private boolean dataRemainingInBlock() {
+    long bufferPos;
+    if (blockPosition >= 0) {
+      bufferPos = blockPosition;
+    } else {
+      bufferPos = bufferOffsetWrtBlockDataData + buffersSize;
+    }
+
+    return bufferPos < length;
+  }
+
+  /**
+   * Check if current buffer had been read till the end.
+   */
+  private boolean bufferEOF() {
+    return allocated && !buffers.get(bufferIndex).hasRemaining();
```

Comment:
Probable Null Pointer exception in buffers.get(...) statement?

Reply:
done